zerver/lib/users.py

```diff
@@ -476,6 +492,204 @@ def format_user_row(
     return result
 
 
+def user_access_restricted_in_realm(target_user: UserProfile) -> bool:
+    if target_user.is_bot:
+        return False
+
+    realm = target_user.realm
+    if realm.can_access_all_users_group.name == SystemGroups.EVERYONE:
+        return False
+
+    return True
+
+
+def check_user_can_access_all_users(acting_user: Optional[UserProfile]) -> bool:
+    if acting_user is None:
+        return True
+
+    if not acting_user.is_guest:
+        return True
+
+    realm = acting_user.realm
+    if is_user_in_group(realm.can_access_all_users_group, acting_user):
+        return True
+
+    return False
+
+
+def check_can_access_user(
+    target_user: UserProfile, user_profile: Optional[UserProfile] = None
+) -> bool:
+    if check_user_can_access_all_users(user_profile) or not user_access_restricted_in_realm(
+        target_user
+    ):
+        return True
+
+    assert user_profile is not None
+
+    if target_user.id == user_profile.id:
+        return True
+
+    # These include Subscription objects for streams as well as group DMs.
+    subscribed_recipient_ids = Subscription.objects.filter(
+        user_profile=user_profile,
+        active=True,
+        recipient__type__in=[Recipient.STREAM, Recipient.HUDDLE],
+    ).values_list("recipient_id", flat=True)
+
+    if Subscription.objects.filter(
+        recipient_id__in=subscribed_recipient_ids, user_profile=target_user, active=True
+    ).exists():
+        return True
+
+    assert user_profile.recipient_id is not None
+    assert target_user.recipient_id is not None
+    direct_message_query = Message.objects.filter(
+        recipient__type=Recipient.PERSONAL, realm=target_user.realm
+    )
+    if direct_message_query.filter(
+        Q(sender_id=target_user.id, recipient_id=user_profile.recipient_id)
+        | Q(recipient_id=target_user.recipient_id, sender_id=user_profile.id)
+    ).exists():
+        return True
+
+    return False
+
+
+def get_user_ids_who_can_access_user(target_user: UserProfile) -> List[int]:
+    realm = target_user.realm
+    if not user_access_restricted_in_realm(target_user):
+        return active_user_ids(realm.id)
+
+    active_non_guest_user_ids_in_realm = active_non_guest_user_ids(realm.id)
+
+    users_in_subscribed_streams_or_huddles_dict = get_subscribers_of_target_user_subscriptions(
+        [target_user]
+    )
+    users_involved_in_dms_dict = get_users_involved_in_dms_with_target_users([target_user], realm)
+
+    user_ids_who_can_access_target_user = (
+        {target_user.id}
+        | set(active_non_guest_user_ids_in_realm)
+        | users_in_subscribed_streams_or_huddles_dict[target_user.id]
+        | users_involved_in_dms_dict[target_user.id]
+    )
+    return list(user_ids_who_can_access_target_user)
```

Comment:
I think the structure for this PR is pretty good. But I expect that `get_users_involved_in_dms_with_target_users` with its queries on the `Message` table, is, I think, far too expensive for a function that will be called often from presence code paths for every user in an organization that has this feature enabled.

I'm not sure what to do about it. I suppose one theory would be that we could work towards creating `Huddle` objects with `Subscription`s for all 1:1 DM conversations (even if we don't do the more complex refactoring that would be required to create corresponding `Recipient` objects and then repoint `Message` objects at them without breaking anything). In other words, doing the first part of the migrations needed to do https://github.com/zulip/zulip/issues/25713.

With that implemented, it would become the case that we at least only need to query the `Subscription`/`Recipient` tables in order to do these checks. That might still be too slow ... but it at least would be a much simpler problem statement to optimize.